### ISSUE TITLE
[Native] Update StyledNativeComponent's ref logic to match StyledComponent

### ISSUE THREAD

**Issue by FakeYou** [styled-components/styled-components#674]
## Version
- 1.4.4  
- react: 16.0.0-alpha.6
- react-native: 0.43.2

## Steps to reproduce
- create a styled react native component eg: `const Container = styled.View`
- render the component

## Expected Behavior
no warnings

## Actual Behavior
A warning stating:
> Stateless function components cannot be given refs. Attempts to access this ref will fail.

I just upgrade my react-native project from 0.42 to 0.43 and since then I got this warning. I've looked into it a little bit and I think it is cause by [this line](https://github.com/styled-components/styled-components/blob/f5c63828540baa44636c78e8287f89838f66d860/src/models/StyledNativeComponent.js#L113). If I look at the `StyleComponent.js` file I can see a check for `innerRef` before settings the `ref` props ([here](https://github.com/styled-components/styled-components/blob/f5c63828540baa44636c78e8287f89838f66d860/src/models/StyledComponent.js#L96))

In [this commit](https://github.com/styled-components/styled-components/commit/f5c63828540baa44636c78e8287f89838f66d860) the check for `innerRef` is removed but I'm not sure why, the description only states 'Fix error'.

To fix this issue I think this change in `StyledNativeComponent.js` shoud work.

```js
    render() {
      const { style, children, innerRef } = this.props // <-- grab innerRef from this.props
      const { generatedStyles } = this.state

      const propsForElement = { ...this.props }
      propsForElement.style = [generatedStyles, style]

      if (innerRef) { // <-- check if innerRef is set before setting the ref prop for propsForElement
        propsForElement.ref = this.generateRef()
        if (isTag(target)) delete propsForElement.innerRef
      }

      return createElement(target, propsForElement, children)
    }
```

**Comment by philpl** [comment]
We'll probably fix this in v2 only, but there's some updating that needs to be done on the native counterpart of the StyledComponent anyway.

**Comment by philpl** [comment]
This should be fixed in #714 since the implementations will match in v2 after merging it

**Comment by dbertella** [comment]
I have the same warning I'm on RN 0.44.0 and I have this when I try to compose from a functional component. ex: 
```
const PrimaryContainer = styled(Container)
...
```

```
// Container.js
export default (props: Props) => <Container {...props} />;
```

**Comment by dbertella** [comment]
I can confirm that switching:
```
// Container.js
export default (props: Props) => <Container {...props} />;
```
to
```
class C extends React.Component {
  props: Props;
  render() {
    return <Container {...props} />;
  }
}

export default C;
```
Will solve the problem.
Any thoughts about it?

**Comment by philpl** [comment]
@dbertella You're probably trying to use `innerRef` to get a ref of that `Container.js` component, but unfortunately React doesn't allow you to get a `ref` to functional components. You should probably create your own `innerRef` prop e.g. `_innerRef` or `domRef`, and pass that to your DOM element.

**Comment by dbertella** [comment]
Hi Phil thanks for the suggestion but it's really weird since I don't
usually use ref anywhere in my code. Can it be that some react native
components I'm thinking at the listview for example rely on that
underneath? It sounds weird to me because it looks like i have this warning
every time I try to compose a new component that way. I will try to put an
example together if you think it's worth.

**Comment by philpl** [comment]
@dbertella Can you open a new issue with the minimal code to reproduce this? Something is probably using the `innerRef` prop though, since I don't see how this could be caused in any other way. However, the version of SC you're using is a factor as well
